File: go/day6/puzzle.go
package day6

import (
	"strings"
)

func SolvePart1(memoryBanks []int) int {
	redistributionCycles := 0

	history := map[string]bool{}

	for !historyContains(history, memoryBanks) {
		addToHistory(history, memoryBanks)

		index := indexOfMax(memoryBanks)

		reallocate(memoryBanks, index)

		redistributionCycles++
	}

	return redistributionCycles
}

func SolvePart2(memoryBanks []int) int {
	// The solution to part 2 is the same as to part 1, the only thing that differs is the initial memory banks
    return SolvePart1(memoryBanks)
}

func reallocate(memoryBanks []int, index int) {
	blockCount := memoryBanks[index]

	// Clear block count
	memoryBanks[index] = 0

	// Reallocate
	for offset := 1; offset <= blockCount; offset++ {
		currentIndex := (index + offset) % len(memoryBanks)

		memoryBanks[currentIndex]++
	}
}

func historyContains(history map[string]bool, memoryBanks []int) bool {
	return history[toString(memoryBanks)]
}

func addToHistory(history map[string]bool, memoryBanks []int) {
	history[toString(memoryBanks)] = true
}

func toString(values []int) string {
	valuesAsStrings := []string{}

	for _, value := range values {
		valuesAsStrings = append(valuesAsStrings, string(value))
	}

	return strings.Join(valuesAsStrings, ",")
}

func indexOfMax(values []int) int {
	index := 0
	max := values[0]

	for i, value := range values {
		if value > max {
			index = i
			max = value
		}
	}

	return index
}

File: 2017/dotnet/src/Day6/Puzzle.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AdventOfCode.Day6
{
    public static class Puzzle
    {   
        public static int SolvePart1(int[] memoryBanks)
        {
            var redistributionCycles = 0;

            var history = new History();

            while (!history.Contains(memoryBanks))
            {
                history.Add(memoryBanks);

                var index = Array.IndexOf(memoryBanks, memoryBanks.Max());

                Reallocate(memoryBanks, index);

                redistributionCycles++;
            }
            
            return redistributionCycles;
        }

        public static int SolvePart2(int[] memoryBanks)
        {
            // The solution to part 2 is the same as to part 1, the only thing that differs is the
            // initial memory banks
            return SolvePart1(memoryBanks);
        }
        
        private static void Reallocate(int[] memoryBanks, int index)
        {
            var blockCount = memoryBanks[index];

            // Clear block count
            memoryBanks[index] = 0;

            // Reallocate
            for (int offset = 1; offset <= blockCount; offset++)
            {
                var currentIndex = (index + offset) % memoryBanks.Length;

                memoryBanks[currentIndex]++;
            }
        }

        private class History
        {
            private readonly List<string> history;

            public History()
            {
                history = new List<string>();
            }

            public void Add(int[] memoryBanks) =>
                history.Add(ToString(memoryBanks));

            public bool Contains(int[] memoryBanks) =>
                history.Contains(ToString(memoryBanks));

            private static string ToString(int[] memoryBanks) =>
                string.Join(",", memoryBanks);
        }
    }
}
